pkg/controllers/routereflector/ctrl_ops.go

```diff
@@ -0,0 +1,295 @@
+
+	if c.topology.IsRouteReflector(string(kubeNode.GetUID()), kubeNode.GetLabels()) {
+		// Node is deleted right now or has some issues, better to remove form RRs
+		if err := c.removeRRStatus(kubeNode); err != nil {
+			log.Errorf("Unable to cleanup label on %s: %s", kubeNode.GetName(), err.Error())
+			return err
+		}
+		log.Infof("Label was removed from node %s", kubeNode.GetName())
+	}
+
+	log.Infof("Time to re-reconcile for node %s", kubeNode.GetName())
+	return c.update(kubeNode)
+}
+
+// update is triggered when a node has beed updated on the cluster
+func (c *ctrl) update(kubeNode *corev1.Node) error {
+	affectedNodes := c.collectNodeInfo(c.topology.GetNodeFilter(kubeNode))
+	log.Debug("Affected nodes: %#v", affectedNodes)
+
+	if err := c.updateNodeLabels(affectedNodes); err != nil {
+		return err
+	} else if err := c.updateBGPTopology(affectedNodes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// revertFailedModifications detects prevoius execution failure and tries to revert the applied change
+func (c *ctrl) revertFailedModifications() error {
+	// Iterating on all known nodes
+	for _, kubeNode := range c.kubeNodes {
+		status, ok := c.routeReflectorsUnderOperation[kubeNode.GetUID()]
+		if !ok {
+			continue
+		}
+
+		rrRevertCounter.WithLabelValues(kubeNode.GetName()).Inc()
+
+		// Node was under operation, better to revert it
+		var err error
+		if status {
+			// Previous operation was add
+			err = c.datastore.RemoveRRStatus(kubeNode, c.findCalicoNode(kubeNode))
+		} else {
+			// Previous operaion was remove
+			err = c.datastore.AddRRStatus(kubeNode, c.findCalicoNode(kubeNode))
+		}
+		if err != nil {
+			log.Errorf("Failed to revert node %s: %s", kubeNode.GetName(), err.Error())
+			return err
+		}
+
+		// Update Kubernetes node
+		if _, err := c.k8sNodeClient.Update(context.Background(), kubeNode, metav1.UpdateOptions{}); err != nil {
+			log.Errorf("Failed to revert update node %s: %s", kubeNode.GetName(), err.Error())
+			return err
+		}
+
+		delete(c.routeReflectorsUnderOperation, kubeNode.GetUID())
+	}
+
+	return nil
+}
+
+// updateNodeLabels calculates topology and updates all the necessary labels in datastore
+func (c *ctrl) updateNodeLabels(affectedNodes map[*corev1.Node]bool) error {
+	// on case of a zone has less nodes then required, count missing
+	missingRouteReflectors := 0
+
+	// Calculate topology
+	for _, status := range c.topology.GetRouteReflectorStatuses(affectedNodes) {
+		// Increment the expected with the number of missing
+		status.ExpectedRRs += missingRouteReflectors
+		log.Infof("Route refletor status in zone(s) %v (actual/expected) = %d/%d", status.Zones, status.ActualRRs, status.ExpectedRRs)
+
+		for _, n := range status.Nodes {
+			n := c.kubeNodes[n.GetUID()]
+
+			if !affectedNodes[n] {
+				continue
+			} else if status.ExpectedRRs == status.ActualRRs {
+				break
+			}
+
+			// Calculate diff to detect scale operation
+			diff := status.ExpectedRRs - status.ActualRRs
+			if diff == 0 {
```

Comment:
I think if we reach this point we're guaranteed that `diff != 0` because of the the if statement above where we `break` if Expected == Actual